sympy/sets/sets.py

```diff
@@ -193,21 +193,19 @@ def complement(self, universe):
 
     def _complement(self, other):
         # this behaves as other - self
-        if isinstance(other, ProductSet):
-            # For each set consider it or it's complement
-            # We need at least one of the sets to be complemented
-            # Consider all 2^n combinations.
-            # We can conveniently represent these options easily using a
-            # ProductSet
-
-            # XXX: this doesn't work if the dimensions of the sets isn't same.
-            # A - B is essentially same as A if B has a different
-            # dimensionality than A
-            switch_sets = ProductSet(*(FiniteSet(o, o - s) for s, o in
-                                     zip(self.sets, other.sets)))
-            product_sets = (ProductSet(*set) for set in switch_sets)
-            # Union of all combinations but this one
-            return Union(*(p for p in product_sets if p != other))
+        if isinstance(self, ProductSet) and isinstance(other, ProductSet):
+            # There can be other ways to represent this but this gives:
+            # (A x B) - (C x D) = ((A - C) x B) U (A x (B - D))
+
+            if len(self.sets) != len(other.sets):
+                raise ValueError("A is not a subset of B")
```

Comment:
There exists complement between product sets even if ![image](https://user-images.githubusercontent.com/34944973/64413764-11fa4c00-d0cd-11e9-8236-60791f215ccb.png). And it's more simple.

Reply:
I guess you mean the empty set. Fair point :)

Reply:
No, I mean that if tuples of different length are considered not equal, the products should be disjoint and the complement can be reduced to `A`

However, if having different length tuples are not considered to be unequal each other even if they have different length, it should stay unevaluated.
It is from the statement in
https://math.stackexchange.com/questions/2122856/endertons-a-mathematical-introduction-to-logic-question-about-n-tuples-in-th

And I get that the statement from the Herbert can only prove the non-associative like 
![image](https://user-images.githubusercontent.com/34944973/64424695-b340cc80-d0e4-11e9-9fb9-4c2327cb058b.png) 
but flattening to the left or to the right 
![image](https://user-images.githubusercontent.com/34944973/64424731-cb185080-d0e4-11e9-8133-5721311d4f41.png) can still be possible.

Probably we've initially got it wrong with associativity if 
![image](https://user-images.githubusercontent.com/34944973/64425077-c7d19480-d0e5-11e9-824f-1017380ca948.png)
doesn't necessarily imply that
![image](https://user-images.githubusercontent.com/34944973/64425168-ffd8d780-d0e5-11e9-9db7-a0f588a259db.png) or ![image](https://user-images.githubusercontent.com/34944973/64425192-0ff0b700-d0e6-11e9-9765-ba31c067fb8b.png)

But I see the implementation here would stay okay, because we also have to make the definition for the tuple consistent with the python object.

Reply:
> No, I mean that if tuples of different length are considered not equal, the products should be disjoint and the complement can be reduced to `A`

Oh of course. If A n B is empty then A - B = A.

I don't think we "got it wrong". We're just using a different definition. I don't see any particular advantage to the recursive definitions.